sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -87,9 +102,12 @@ class _BinMapper(BaseEstimator, TransformerMixin):
     Parameters
     ----------
     max_bins : int, optional (default=256)
-        The maximum number of bins to use. If for a given feature the number of
-        unique values is less than ``max_bins``, then those unique values
-        will be used to compute the bin thresholds, instead of the quantiles.
+        The maximum number of bins to use (including the bin for missing
+        values). If for a given feature the number of unique values is less
+        than ``max_bins - 1``, then those unique values will be used to
+        compute the bin thresholds, instead of the quantiles. The first bin
+        is always reserved for missing values, so the number of bins used
+        for non-missing values is actually ``max_bins - 1``.
```

Comment:
an alternative is of course to have the actual number of bins to be `max_bins + 1` (for the missing values), and limit the `max_bins` to 254 instead. Not sure which one makes more sense.

Reply:
That would break when we encode bins in uint32 instead of uint8

lightgbm uses the last bin for missing values, but the last bin may vary. If there are 6 bins for non-missing values, the last bin is the 7th one, not the 256th one. That requires every function to know which is the last bin on every feature.

As mentionned in the header of this PR, I can't think of any reason why this is more convenient

Reply:
I didn't mean to have it at the end, I meant to have `max_bins` to be:

```
        The maximum number of bins to use for values other than ``NaN``
        If for a given feature the number of unique values is less
        than ``max_bins``, then those unique values will be used to
        compute the bin thresholds, instead of the quantiles. The maximum
        supported value is `255`.
```

Reply:
Now max_bins isn't "max bins" anymore.

I don't mind too much, it just shifts the discrepancy somewhere else IMO.

@ogrisel WDYT?

Reply:
I've tried implementing your proposed version, where max_bins is the number of bins for non-missing values.

It didn't make the code simpler. Most of the tests still need to be adapted. It even has the unwanted effect that `max_bins` passed to `HistGradientBoosting***` doesn't have the same meaning as other `max_bins` parameters, typically as passed to `HistogramBuilder` (where max_bins must absolutely include missing values).

That being said, I agree that the docstring might be a bit confusing, and that ideally users shouldn't have to worry about the bin reserved missing values.

Feedback welcome.

Reply:
I think we submitted our comments about the same time @ogrisel 	, did you get a chance to consider https://github.com/scikit-learn/scikit-learn/pull/13911/files#r297715531 ?